jerry-core/ecma/builtin-objects/ecma-builtin-json.cpp

```diff
@@ -0,0 +1,1815 @@
+{
+  JERRY_ASSERT (reviver_p);
+  JERRY_ASSERT (holder_p);
+  JERRY_ASSERT (name_p);
+
+  ecma_completion_value_t completion_value = ecma_op_object_get (holder_p, name_p);
+
+  if (!ecma_is_completion_value_normal (completion_value))
+  {
+    return completion_value;
+  }
+
+  ecma_value_t value = ecma_copy_value (ecma_get_completion_value_value (completion_value), true);
+  ecma_free_completion_value (completion_value);
+  completion_value = ecma_make_empty_completion_value ();
+
+  if (ecma_is_value_object (value))
+  {
+    ecma_object_t *object_p = ecma_get_object_from_value (value);
+    ecma_property_t *properties_p = ecma_get_property_list (object_p);
+    uint32_t no_properties = 0;
+    ecma_property_t *property_p;
+
+    /*
+     * The following algorithm works with arrays and objects as well.
+     */
+
+    for (property_p = properties_p;
+         property_p != NULL;
+         property_p = ECMA_GET_POINTER (ecma_property_t, property_p->next_property_p))
+    {
+      /*
+       * All properties must be named properties, since we constructed them.
+       */
+      JERRY_ASSERT (property_p->type == ECMA_PROPERTY_NAMEDDATA
+                    || property_p->type == ECMA_PROPERTY_INTERNAL);
+
+      if (property_p->type == ECMA_PROPERTY_NAMEDDATA
+          && property_p->u.named_data_property.enumerable)
+      {
+        no_properties++;
+      }
+    }
+
+    if (no_properties > 0)
+    {
+      size_t buffer_size = sizeof (ecma_string_t*) * (no_properties);
+      ecma_string_t **property_names_p = (ecma_string_t**) mem_heap_alloc_block (buffer_size,
+                                                                                 MEM_HEAP_ALLOC_SHORT_TERM);
+
+      if (property_names_p == NULL)
+      {
+        jerry_fatal (ERR_OUT_OF_MEMORY);
+      }
+
+      uint32_t property_index = no_properties;
+      for (property_p = properties_p;
+           property_p != NULL;
+           property_p = ECMA_GET_POINTER (ecma_property_t, property_p->next_property_p))
+      {
+        if (property_p->type == ECMA_PROPERTY_NAMEDDATA
+            && property_p->u.named_data_property.enumerable)
+        {
+          ecma_string_t *property_name_p = ECMA_GET_NON_NULL_POINTER (ecma_string_t,
+                                                                      property_p->u.named_data_property.name_p);
+         /*
+          * The property list must be in creation order.
+          */
+          property_index--;
+          property_names_p[property_index] = ecma_copy_or_ref_ecma_string (property_name_p);
+        }
+      }
+
+      JERRY_ASSERT (property_index == 0);
+
+      for (property_index = 0; property_index < no_properties; property_index++)
+      {
+        completion_value = ecma_builtin_json_walk (reviver_p,
+                                                   object_p,
+                                                   property_names_p[property_index]);
+        if (!ecma_is_completion_value_normal (completion_value))
```

Comment:
There are some places in the function that still do not follow the `ECMA_TRY_CATCH` style.
I've tried to update them. Could you, please, check and consider applying the following patch?

```
diff --git a/jerry-core/ecma/builtin-objects/ecma-builtin-json.cpp b/jerry-core/ecma/builtin-objects/ecma-builtin-json.cpp
index 510dc1b..d34717d 100644
--- a/jerry-core/ecma/builtin-objects/ecma-builtin-json.cpp
+++ b/jerry-core/ecma/builtin-objects/ecma-builtin-json.cpp
@@ -608,40 +608,33 @@ ecma_builtin_json_walk (ecma_object_t *reviver_p, /**< reviver function */
   JERRY_ASSERT (holder_p);
   JERRY_ASSERT (name_p);
 
-  ecma_completion_value_t completion_value = ecma_op_object_get (holder_p, name_p);
-
-  if (!ecma_is_completion_value_normal (completion_value))
-  {
-    return completion_value;
-  }
+  ecma_completion_value_t ret_value = ecma_make_empty_completion_value ();
 
-  ecma_value_t value = ecma_copy_value (ecma_get_completion_value_value (completion_value), true);
-  ecma_free_completion_value (completion_value);
-  completion_value = ecma_make_empty_completion_value ();
+  ECMA_TRY_CATCH (value_get,
+                  ecma_op_object_get (holder_p, name_p),
+                  ret_value);
 
-  if (ecma_is_value_object (value))
+  if (ecma_is_value_object (value_get))
   {
-    ecma_object_t *object_p = ecma_get_object_from_value (value);
-    ecma_property_t *properties_p = ecma_get_property_list (object_p);
+    ecma_object_t *object_p = ecma_get_object_from_value (value_get);
+
     uint32_t no_properties = 0;
-    ecma_property_t *property_p;
 
     /*
      * The following algorithm works with arrays and objects as well.
      */
-
-    for (property_p = properties_p;
+    for (ecma_property_t *property_p = ecma_get_property_list (object_p);
          property_p != NULL;
          property_p = ECMA_GET_POINTER (ecma_property_t, property_p->next_property_p))
     {
       /*
-       * All properties must be named properties, since we constructed them.
+       * All properties must be named data or internal properties, since we constructed them.
        */
       JERRY_ASSERT (property_p->type == ECMA_PROPERTY_NAMEDDATA
                     || property_p->type == ECMA_PROPERTY_INTERNAL);
 
       if (property_p->type == ECMA_PROPERTY_NAMEDDATA
-          && property_p->u.named_data_property.enumerable)
+          && ecma_is_property_enumerable (property_p))
       {
         no_properties++;
       }
@@ -652,12 +645,12 @@ ecma_builtin_json_walk (ecma_object_t *reviver_p, /**< reviver function */
       MEM_DEFINE_LOCAL_ARRAY (property_names_p, no_properties, ecma_string_t *);
 
       uint32_t property_index = no_properties;
-      for (property_p = properties_p;
+      for (ecma_property_t *property_p = ecma_get_property_list (object_p);
            property_p != NULL;
            property_p = ECMA_GET_POINTER (ecma_property_t, property_p->next_property_p))
       {
         if (property_p->type == ECMA_PROPERTY_NAMEDDATA
-            && property_p->u.named_data_property.enumerable)
+            && ecma_is_property_enumerable (property_p))
         {
           ecma_string_t *property_name_p = ECMA_GET_NON_NULL_POINTER (ecma_string_t,
                                                                       property_p->u.named_data_property.name_p);
@@ -670,44 +663,37 @@ ecma_builtin_json_walk (ecma_object_t *reviver_p, /**< reviver function */
       }
 
       JERRY_ASSERT (property_index == 0);
-      completion_value = ecma_make_empty_completion_value ();
 
-      for (property_index = 0; property_index < no_properties; property_index++)
+      for (property_index = 0;
+           property_index < no_properties && ecma_is_completion_value_empty (ret_value);
+           property_index++)
       {
-        ECMA_TRY_CATCH (value,
+        ECMA_TRY_CATCH (value_walk,
                         ecma_builtin_json_walk (reviver_p,
                                                 object_p,
                                                 property_names_p[property_index]),
-                        completion_value);
+                        ret_value);
 
         /*
          * We cannot optimize this function since any members
          * can be changed (deleted) by the reviver function.
          */
-
-        if (ecma_is_value_undefined (value))
+        if (ecma_is_value_undefined (value_walk))
         {
-          ecma_completion_value_t comp_val = ecma_op_general_object_delete (object_p,
-                                                                            property_names_p[property_index],
-                                                                            false);
-          ecma_free_completion_value (comp_val);
+          ecma_completion_value_t delete_val = ecma_op_general_object_delete (object_p,
+                                                                              property_names_p[property_index],
+                                                                              false);
+          JERRY_ASSERT (ecma_is_completion_value_normal_true (delete_val)
+                        || ecma_is_completion_value_normal_false (delete_val));
         }
         else
         {
           ecma_builtin_json_define_value_property (object_p,
                                                    property_names_p[property_index],
-                                                   value);
+                                                   value_walk);
         }
 
-        ecma_free_completion_value (completion_value);
-        completion_value = ecma_make_empty_completion_value ();
-
-        ECMA_FINALIZE (value);
-
-        if (!ecma_is_completion_value_empty (completion_value))
-        {
-          break;
-        }
+        ECMA_FINALIZE (value_walk);
       }
 
       for (uint32_t i = 0; i < no_properties; i++)
@@ -719,24 +705,29 @@ ecma_builtin_json_walk (ecma_object_t *reviver_p, /**< reviver function */
     }
   }
 
-  /*
-   * Completition value can be an anything including exceptions.
-   */
-  if (ecma_is_completion_value_empty (completion_value))
+  if (ecma_is_completion_value_empty (ret_value))
   {
     ecma_value_t arguments_list[2];
 
     arguments_list[0] = ecma_make_string_value (name_p);
-    arguments_list[1] = value;
-
-    completion_value = ecma_op_function_call (reviver_p,
-                                              ecma_make_object_value (holder_p),
-                                              arguments_list,
-                                              2);
+    arguments_list[1] = value_get;
+
+   /*
+    * The completion value can be anything including exceptions.
+    */
+    ret_value = ecma_op_function_call (reviver_p,
+                                       ecma_make_object_value (holder_p),
+                                       arguments_list,
+                                       2);
+  }
+  else
+  {
+    JERRY_ASSERT (ecma_is_completion_value_throw (ret_value));
   }
 
-  ecma_free_value (value, true);
-  return completion_value;
+  ECMA_FINALIZE (value_get);
+
+  return ret_value;
 } /* ecma_builtin_json_walk */
 
 /**
```